### ISSUE TITLE
An issue with the estimation of the support in reid function.

### ISSUE THREAD

**Issue by lionelkusch** [mind-inria/hidimstat#266]
By changing the data in the tests, I find that `size_support` is not well estimated. I can't fix it because the tolerance is used in the calculation of the `size_support` and in the Lasso.

_Originally posted by @lionelkusch in https://github.com/mind-inria/hidimstat/pull/265#discussion_r2105193505_

**Comment by bthirion** [comment]
You mean that we need two different tolerance values ? 
However, beware that when a lasso-type yields very small values, it actually means that the value should be 0 and the estimator did not converge properly.

**Comment by jpaillard** [comment]
Agreed with @bthirion . 
In the [Note of the Lasso doc](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html#lasso) doc, they write that the optimization (hidden in `Refit_CV`) stops when the updates are smaller than ` tolerance * coefficients_.max()`. To estimate the support it then makes sense to use the same criteria. 

For consistency with the sklearn API, I suggest renaming this parameter `tol`

**Comment by lionelkusch** [comment]
I don't see the link between that optimisation criterion of Lasso and the support size. 
Can you give me a hint or a reference?

**Comment by jpaillard** [comment]
You get the support size by counting "non-zero coefficients". 
As pointed out by Bertrand, the actual value of the "null coefficient" might be numerically small but not exactly zero because of the optimization tolerance parameter -- the algo will stop performing updates when they are smaller than `tolerance * coefficients_.max()` (see notes Lasso docstring). So we only consider coefficients above this as part of the support.

**Comment by lionelkusch** [comment]
For the tests on reid, I, for the moment, fix them by setting a good seed, but later this high variability induced by the seed should be removed.

_Originally posted by @lionelkusch in https://github.com/mind-inria/hidimstat/issues/265#issuecomment-2966578955_

**Comment by bthirion** [comment]
Thx for reporting.  It turns out that your expectations are excessive: you want to accurately estimate a full covariance matrix on a coefficient-by-coefficient matrix. For that you should use the Ledoit-Wolf estimator.
Reid is meant to estimate the scalar variance of the noise, not the full covariance matrix.

**Comment by lionelkusch** [comment]
The output of the function reid is the standard deviation of the noise when there is no temporal axis. 
In the case of the temporal axis, there is an estimate of the covariance matrix of the noise. I assume that it's the variation in the noise over time.
From the tests, the estimation of the covariance matrix of the noise is not robust. I don't think that the Ledoit-Wolf estimator is a good method for evaluating the variance of the noise. From my understanding, it's more used in the covariance matrix of the signal.

**Comment by bthirion** [comment]
Sure,  but you cannot expect a covariance estimated on limited data to be accurate.

**Comment by lionelkusch** [comment]
The value of the diagonal is always the same. 
I compared this value to different seeds in the experience 1 of `test_group_reid`.
I got the following results:
|  expected values |  estimated values |
| ------------------------  | -------------------------- |
|0.2725524165152086| 0.3087268658868705|  
|0.2109871530439731| 0.2819540812369865|  
|0.2220124601817509| 0.2761317527678781|  
|0.1811807858508895| 0.5643770530206996|  
|0.1758223885104770| 1.6372040161404070|  
|0.1348328499801373| 0.2409761030861925|  
|0.2324349919636746| 0.5088612091885980|  
|0.1904840165803888| 0.2263553243246939|  
|0.1517071402613948| 0.2998269114293044|  
|0.2138081200546301| 0.2778642052304738|  
|0.1805317640106714| 0.2543458303001847|  
|0.1488379475067388| 0.9583748273020606|  
|0.2286939223079950| 0.4423802355128778|  
|0.1447343597115248| 0.8457074994967290|  
|0.0990080974217503| 0.3036686453371481|  
|0.2836830728790790| 0.7750570318114842|  
|0.1549449451036520| 0.2280851517132253|  
|0.2005786242400802| 0.4597541279555092|  
|0.1700323450413110| 0.6626485784329070|  
|0.2275921212110201| 0.2396646275093938| 

We can see the difference can be very small (0.02/1%) at big (1.0/1000%).
Do you think that the size of the dataset (100) is not large enough?

**Comment by bthirion** [comment]
I'm not sure I understand: what quantity do these numbers represent ? The noise (scalar) variance ? If yes, this is a bit worrying.

**Comment by lionelkusch** [comment]
yes, it's the noise variance.

**Comment by bthirion** [comment]
To move fw, please provide a concise script that yields this value.  Please try to use the code of a given branch rather than copy-pasting it.

**Comment by lionelkusch** [comment]
I increased the number of times and the number of samples.
It seems that I need to increase the time and the number of samples for having more stable tests.
Do you have some orders of magnitude for samples, features, time and SNR for having reasonable tests?

**Comment by bthirion** [comment]
For the moment yes. I think that there's bug in the AR case, because the correlation between features and across time is difficult to handle. However, I think it's and edge case, so let's merge and open an issue to keep track of that.

**Comment by lionelkusch** [comment]
This is already the issue for tracking it.
I need to make a PR for improving the tests for REID because it has an impact on Desparsified Lasso and EnCluDL.

**Comment by bthirion** [comment]
OK, my suggestion is to indeed add a test that works OK-ish, then leava an open issue for the temporal autocorrelation case (which requires to go into the maths).
